### PR TITLE
[feature] data synchronization function

### PR DESCRIPTION
**Pull request by wenhemin**
please see this issue #1856

### CODE REVIEW

**Comment by dailidong** [comment]
pstmt and resultset need close

https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=1881&resolved=false&types=BUG

**Comment by wenhemin** [comment]
> 1，datax or sqoop is data import or export, what's mean ETL task naming ?
> 2，suggest remove class author，because other not

1，The front end is named ETL, and the inside is implemented by Datax. It can be replaced by other implementations.
2，fixed

**Comment by wenhemin** [comment]
> pstmt and resultset need close
> 
> https://sonarcloud.io/project/issues?id=apache-dolphinscheduler&pullRequest=1881&resolved=false&types=BUG

fixed

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    /**
+     * data source type，eg  MYSQL, POSTGRES ...
+     */
+    private String dsType;
+
+    /**
+     * datasource id
+     */
+    private int datasource;
```
Comment (Yeleights) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java]: datasource -> dataSource

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    /**
+     * data target type，eg  MYSQL, POSTGRES ...
+     */
+    private String dtType;
+
+    /**
+     * datatarget id
+     */
+    private int datatarget;
```
Comment (Yeleights) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java]: datatarget -> dataTarget

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        return jobSpeedRecord;
+    }
+
+    public void setJobSpeedRecord(int jobSpeedRecord) {
+        this.jobSpeedRecord = jobSpeedRecord;
+    }
+
+    @Override
+    public boolean checkParameters() {
+        if (!(datasource != 0 && StringUtils.isNotEmpty(dsType) && StringUtils.isNotEmpty(sql))) {
```
Comment (Yeleights) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java]: Whether these two if condition are considered to be combined into one？

**Comment by Yeleights** [comment]
> > 1，datax or sqoop is data import or export, what's mean ETL task naming ?
> > 2，suggest remove class author，because other not
> 
> 1，The front end is named ETL, and the inside is implemented by Datax. It can be replaced by other implementations.
> 2，fixed

data sync as this task name is more suitable than etl? 
Datax or sqoop should only be responsible for the import and export of data. Although it can take part in data process, the main data process should be implemented by hive-ql or spark.

**Comment by Jave-Chen** [comment]
New files UT coverage need add files at pom.xml at root. Like this:
 <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>${maven-surefire-plugin.version}</version>
                <configuration>
                    <includes>
                        <include>**/common/utils/*.java</include>
                        <include>**/common/threadutils/*.java</include>
                        <include>**/common/graph/*.java</include>
                        <include>**/common/queue/*.java</include>
                        <include>**/api/utils/CheckUtilsTest.java</include>
                        <include>**/api/utils/FileUtilsTest.java</include>
                        <include>**/api/utils/exportprocess/DataSourceParamTest.java</include>
                        <include>**/api/utils/exportprocess/DependentParamTest.java</include>
                        <include>**/api/enums/*.java</include>
                        <include>**/api/service/AccessTokenServiceTest.java</include>
                        <include>**/api/service/QueueServiceTest.java</include>
                        <include>**/api/service/MonitorServiceTest.java</include>
                        <include>**/api/service/SessionServiceTest.java</include>
                        <include>**/api/service/UsersServiceTest.java</include>
                        <include>**/api/service/TenantServiceTest.java</include>
                        <include>**/api/service/WorkerGroupServiceTest.java</include>
                        <include>**/api/service/AlertGroupServiceTest.java</include>
                        <include>**/api/service/ProjectServiceTest.java</include>
                        <include>**/api/service/ProcessDefinitionServiceTest.java</include>
                        <include>**/api/service/UdfFuncServiceTest.java</include>
                        <include>**/api/service/ResourcesServiceTest.java</include>
                        <include>**/api/service/ExecutorService2Test.java</include>
                        <include>**/api/service/BaseServiceTest.java</include>
                        <include>**/api/service/BaseDAGServiceTest.java</include>
                        <include>**/alert/utils/ExcelUtilsTest.java</include>
                        <include>**/alert/utils/FuncUtilsTest.java</include>
                        <include>**/alert/utils/JSONUtilsTest.java</include>
                        <include>**/alert/utils/PropertyUtilsTest.java</include>
                        <include>**/server/utils/SparkArgsUtilsTest.java</include>
                        <include>**/server/utils/FlinkArgsUtilsTest.java</include>
                        <include>**/server/utils/ParamUtilsTest.java</include>
                        <include>**/server/utils/ScheduleUtilsTest.java</include>
                        <include>**/server/master/MasterExecThreadTest.java</include>
                        <include>**/dao/mapper/AccessTokenMapperTest.java</include>
                        <include>**/dao/mapper/AlertGroupMapperTest.java</include>
                        <include>**/dao/mapper/AlertMapperTest.java</include>
                        <include>**/dao/mapper/CommandMapperTest.java</include>
                        <include>**/alert/template/AlertTemplateFactoryTest.java</include>
                        <include>**/alert/template/impl/DefaultHTMLTemplateTest.java</include>
                        
                        <include>**/server/worker/task/AbstractTask.java</include>
                    </includes>
                    
                </configuration>
            </plugin>

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java (hunk 1 above)
Comment (wenhemin) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java]: thx, let me fix it

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java (hunk 2 above)
Comment (wenhemin) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java]: thx, let me fix it

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java (hunk 3 above)
Comment (wenhemin) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/etl/EtlParameters.java]: thx, let me fix it

**Comment by wenhemin** [comment]
> > > 1，datax or sqoop is data import or export, what's mean ETL task naming ?
> > > 2，suggest remove class author，because other not
> > 
> > 
> > 1，The front end is named ETL, and the inside is implemented by Datax. It can be replaced by other implementations.
> > 2，fixed
> 
> data sync as this task name is more suitable than etl?
> Datax or sqoop should only be responsible for the import and export of data. Although it can take part in data process, the main data process should be implemented by hive-ql or spark.

I think，A better name is DataX

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/DataxUtilsTest.java [hunk 4]
```diff
@@ -84,9 +99,10 @@ public void testConvertKeywordsColumns() throws Exception {
      */
     @Test
     public void testDoConvertKeywordsColumn() throws Exception {
-        String fromColumn = " \"`select`\" ";
-        String targetColumn = "`select`";
-
-        assertEquals(DataxUtils.doConvertKeywordsColumn(DbType.MYSQL, fromColumn), targetColumn);
+        assertEquals(DataxUtils.doConvertKeywordsColumn(DbType.MYSQL, " \"`select`\" "), "`select`");
```
Comment (Yeleights) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/DataxUtilsTest.java]: assertEquals first param is expected value, sceond param is actual value, please swap there params.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/DataxUtilsTest.java [hunk 5]
```diff
@@ -84,9 +99,10 @@ public void testConvertKeywordsColumns() throws Exception {
...
+        assertEquals(DataxUtils.doConvertKeywordsColumn(DbType.POSTGRESQL, " \"`select`\" "), "\"select\"");
```
Comment (Yeleights) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/DataxUtilsTest.java]: assertEquals first param is expected value, sceond param is actual value, please swap there params, from 102-105 lines.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+            Method method = DataxTask.class.getDeclaredMethod("parsingSqlColumnNames", DbType.class, DbType.class, BaseDataSource.class, String.class);
+            method.setAccessible(true);
+            String[] columns = (String[]) method.invoke(dataxTask, DbType.MYSQL, DbType.MYSQL, dataSource, "select 1 as a, 2 as `table` from dual");
+
+            Assert.assertNotNull(columns);
+
+            Assert.assertTrue(columns.length == 2);
+
+            Assert.assertEquals(Arrays.toString(columns), "[`a`, `table`]");
```
Comment (Yeleights) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java]: assertEquals first param is expected value, sceond param is actual value, please swap there params.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        try {
+            Method method = DataxTask.class.getDeclaredMethod("tryGrammaticalAnalysisSqlColumnNames", DbType.class, String.class);
+            method.setAccessible(true);
+            String[] columns = (String[]) method.invoke(dataxTask, DbType.MYSQL, "select t1.a, t1.b from test t1 union all select a, t2.b from (select a, b from test) t2");
+
+            Assert.assertNotNull(columns);
+
+            Assert.assertTrue(columns.length == 2);
+
+            Assert.assertEquals(Arrays.toString(columns), "[a, b]");
```
Comment (Yeleights) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java]: assertEquals first param is expected value, sceond param is actual value, please swap there params.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+            method.setAccessible(true);
+            List<JSONObject> contentList = (List<JSONObject>) method.invoke(dataxTask, null);
+            Assert.assertNotNull(contentList);
+
+            JSONObject content = contentList.get(0);
+            JSONObject reader = (JSONObject) content.get("reader");
+            Assert.assertNotNull(reader);
+
+            String readerPluginName = (String) reader.get("name");
+            Assert.assertEquals(readerPluginName, DataxUtils.DATAX_READER_PLUGIN_MYSQL);
```
Comment (Yeleights) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java]: assertEquals first param is expected value, sceond param is actual value, please swap there params.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+
+            JSONObject writer = (JSONObject) content.get("writer");
+            Assert.assertNotNull(writer);
+
+            String writerPluginName = (String) writer.get("name");
+            Assert.assertEquals(writerPluginName, DataxUtils.DATAX_WRITER_PLUGIN_MYSQL);
```
Comment (Yeleights) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java]: assertEquals first param is expected value, sceond param is actual value, please swap there params.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+    }
+
+    /**
+     * Method: notNull(Object obj, String message)
+     */
+    @Test
+    public void testNotNull()
+        throws Exception {
+        try {
+            Method method = DataxTask.class.getDeclaredMethod("notNull", Object.class, String.class);
```
Comment (Yeleights) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/worker/task/datax/DataxTaskTest.java]: this Test no assert?

**Comment by dailidong** [comment]
Come on

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/DataxUtilsTest.java (hunk 4 above)
Comment (wenhemin) [reply dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/DataxUtilsTest.java]: thx, I fixed

**Comment by wenhemin** [comment]
UT bug is fixed

**Comment by samz406** [comment]
test

**Pull request merged**